### PR TITLE
Support bluetooth connections

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Hold | [Ticket](https://pi-top.atlassian.net/browse/SWE-186) |

#### Main changes

Support bluetooth connections by adding a GATT server with a service and several characteristics to interact with the frontend app.

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
- Download deb
- Make sure that the unstable apt repository is installed. You can install it by running:
```
sudo apt update
sudo apt install pi-top-os-unstable-apt-source
# Run again to update source
sudo apt update
```
- Install deb artifact by running `sudo apt install ./<file>`. It should automatically fetch the new dependencies from the unstable repository.
- Restart the further-link and bluetooth services
```
sudo systemctl daemon-reload
sudo systemctl restart further-link
sudo systemctl restart bluetooth
```
- Install miniscreen app using the deb file from [this PR](https://github.com/pi-top/pi-top-4-Miniscreen/pull/318).
- Enable pairing mode from Overview page of miniscreen app
- Use [this frontend branch](https://github.com/pi-top/further-frontend/pull/1847) for testing.

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: .github/workflows/deb-build.yml [hunk 1]
```diff
@@ -3,8 +3,6 @@ name: Debian Package Build
 on:
   pull_request:
   push:
-    branches:
-      - master
```
Comment (angusjfw) [inline .github/workflows/deb-build.yml:7]: probably can remove this and create a pr early

File: further_link/__main__.py [hunk 2]
```diff
@@ -25,8 +41,70 @@ def port():
...
+def _version():
+    return dumps({"version": __version__})
+
+
+async def status(_):
+    return web.Response(text=_status())
+
+
+async def version(_):
+    return web.Response(text=_version())
```
Comment (angusjfw) [inline further_link/__main__.py:57]: i think move this to something like further_link.endpoint.status

File: further_link/__main__.py [hunk 3]
```diff
@@ -65,18 +137,34 @@ async def version(_):
...
     vnc.create_ssl_certificate()
-    return web.run_app(
-        create_app(),
-        port=port(),
-        ssl_context=ssl_context(),
-        # Default handle_signals=True will ignore sigterm signals whilst there
-        # are requests that are not complete. This isn't appropriate for our
-        # indefinitely running websockets and can cause device shutdown to hang
-        handle_signals=False,
-    )
```
Comment (angusjfw) [inline further_link/__main__.py:79]: maybe we can keep this sync way of starting the web app and just before it start the bluetooth app using the make_sync decorator, or calling create_bt_app from inside create_app.

whatever ends up simpler is basically the way to go. probably best to avoid the tcpsite stuff because it's sort of an itnernal api and if we can avoid the make_sync decorator that would also be nice as it looks a bit confusing

File: further_link/endpoint/upload.py [hunk 4]
```diff
@@ -47,3 +53,31 @@ async def upload(request):
     return web.Response(
         text=json.dumps({"success": True, "fetched_urls": fetched_urls})
     )
+
+
+async def bluetooth_upload(
+    device, uuid: str, message: bytearray, characteristic_to_report_on: str
+):
+    device.write_value("IN_PROGRESS", characteristic_to_report_on)
```
Comment (angusjfw) [inline further_link/endpoint/upload.py:61]: keep this only if necessary, would be nice to remove if poss. if we do keep it maybe it could be a json for consistency.

File: further_link/__main__.py [hunk 5]
```diff
@@ -25,8 +41,70 @@ def port():
...
+        gatt_config = GattConfig("Further", FURTHER_GATT_CONFIG)
+        gatt_config.register_read_handler(
+            PT_SERVICE_UUID, PT_STATUS_CHARACTERISTIC_UUID, _status
+        )
+        gatt_config.register_read_handler(
+            PT_SERVICE_UUID, PT_VERSION_CHARACTERISTIC_UUID, _version
+        )
+        gatt_config.register_write_handler(
+            PT_SERVICE_UUID, PT_APT_VERSION_CHARACTERISTIC_UUID, apt_version_bt
+        )
```
Comment (angusjfw) [inline further_link/__main__.py:82]: it might be more consistent to use two characteristics for this, so there is the same pattern for all read/write response things

File: further_link/util/bluetooth/gatt.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            "Properties": (
+                GATTCharacteristicProperties.read | GATTCharacteristicProperties.notify
+            ),
+            "Permissions": (GATTAttributePermissions.readable),
+            "Value": None,
+        },
+        PT_APT_VERSION_CHARACTERISTIC_UUID: {
+            "Properties": (
+                GATTCharacteristicProperties.write
+                | GATTCharacteristicProperties.read
```
Comment (angusjfw) [inline further_link/util/bluetooth/gatt.py:37]: split into two, one just write one just read

File: further_link/util/bluetooth/device.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    return bytearray(str(value), "utf-8")
+
+
+def decode_value(value) -> str:
+    if not isinstance(value, str):
+        value = str(value, "utf-8")
+    return value
+
+
+class BluetoothDevice:
```
Comment (angusjfw) [inline further_link/util/bluetooth/device.py:21]: i think name BluetoothServer would be clearer that there is only ever one of these that handles the different client device connections

File: further_link/util/bluetooth/messages.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
+from typing import Optional
+
+
+class ChunkedMessageFormat:
+    START = b"START-"
+    STOP = b"STOP-"
```
Comment (angusjfw) [inline further_link/util/bluetooth/messages.py:6]: it would be nice to send data with the start and stop messages - maybe if they were 4 bytes long this would be made easier - the first 4 bytes are always a header and the reset is the data

File: setup.cfg [hunk 9]
```diff
@@ -49,7 +51,9 @@ test =
     pytest-cov
     pytest-timeout
     pytest-aiohttp
+    pytest-mock
     aioresponses
+    pytest-rerunfailures
```
Comment (angusjfw) [inline setup.cfg:56]: hopefully don't need rerunfailures

File: further_link/util/bluetooth/utils.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+import re
+from typing import Dict
+
+
+def bytearray_to_dict(message: bytearray) -> Dict:
+    message_str = message.decode()
+
+    # remove trailing commas
+    message_str = re.sub(",[ \t\r\n]+}", "}", message_str)
+    message_str = re.sub(",[ \t\r\n]+]", "]", message_str)
```
Comment (angusjfw) [inline further_link/util/bluetooth/utils.py:11]: we shouldn't really need this - the clients should send valid json

**Comment by angusjfw** [comment]
We discussed going with a message id approach rather than buffering. We'll drop support for unchunked messages. So chunked messages have the format:
[2-bytes id][3-bytes end chunk index][3-bytes chunk index][data]

receiving logic is like:
- if there's no open message with the id create a new one
- if there is an open message with the id use that one
- append the chunk (for now if it's not the next expected index that's an error, could start skipping these messages)
- if the chunk index is the end chunk index close and handle the message

File: further_link/util/bluetooth/server.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            self.config.tree.get(self._get_service_uuid(characteristic.uuid), {})
+            .get(characteristic.uuid, {})
+            .get("ReadHandler")
+        )
+        if callback and callable(callback):
+            logging.debug(f"Executing ReadHandler callback for {characteristic.uuid}")
+            try:
+                value = str(callback())
+                # TODO: handle long messages using ChunkedMessage
+                self.write_value(value, characteristic.uuid)
```
Comment (angusjfw) [inline further_link/util/bluetooth/server.py:107]: let's try not writing here - we're not expecting people to be subscribed to notifs as well as reading so it shouldn't be necessary, just return the callback value

File: further_link/util/bluetooth/server.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        if ChunkedMessage.is_start_message(value):
+            self._partial_messages[characteristic.uuid] = ChunkedMessage()
+
+        if self._partial_messages.get(characteristic.uuid):
+            self._partial_messages[characteristic.uuid].append(value)
+        else:
+            should_execute_callback = True
+
+        # update characteristic value in all cases
+        self.write_value(value, characteristic.uuid)
```
Comment (angusjfw) [inline further_link/util/bluetooth/server.py:138]: we definitely shouldn't need this one - you shouldn't be able to read/subscribe to a write characteristic in our model, and it might be slowing things down by writing all the incoming messages to the system bluetooth device. so just call the callback

File: further_link/endpoint/run.py [hunk 13]
```diff
@@ -150,6 +195,63 @@ async def on_display_activity(connection_details: VncConnectionDetails):
...
+        return
+
+    global bt_run_manager
+    if bt_run_manager.get(client_uuid) is None:
+
+        async def send_func(message):
+            logging.debug(f"Sending: {message[0:120]}")
+            await device.write_value(message, characteristic_to_report_on)
+
+        bt_run_manager[client_uuid] = RunManager(send_func, user=None, pty=True)
```
Comment (angusjfw) [inline further_link/endpoint/run.py:225]: can we not pass the user and pty options somehow? forget where the conversation went on that. it's not essential right now since they're not actually configurable on the FE yet, but I think the user one may be useful in future

**Review by angusjfw — APPROVED** [review]
